### PR TITLE
Subscription execution hook

### LINKED ISSUE #6: pallet `on_initialize` hook
This pallet `on_initialize` hook must find all subscription that should be executed this block and execute them

### PR DESCRIPTION
**Pull request by gdnathan**
closes #6

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+	use super::*;
+	use frame_support::pallet_prelude::*;
+	use frame_system::pallet_prelude::*;
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		/// Because this pallet emits events, it depends on the runtime's definition of an event.
+		type Event: From<Event<Self>> + IsType<<Self as frame_system::Config>::Event>;
+		/// The trait to manage funds
+		type Currency: Currency<Self::AccountId> + ReservableCurrency<Self::AccountId>;
```
Comment (tdelabro) [inline src/lib.rs:35]: Do we need reservable in this pallet ?
Are we reserving funds ?

File: src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+	}
+
+	#[pallet::hooks]
+	impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {
+		fn on_initialize(n: T::BlockNumber) -> Weight {
+			let _weight = 0;
+
+			let mut subs = unwrap_or_return!(Subscriptions::<T>::get(), 0); // 1 read
+			let tmp = subs.clone();
+			let (exec_block, (_, _)) = unwrap_or_return!(tmp.last(), 0); // 1 unwrap
```
Comment (tdelabro) [inline src/lib.rs:96]: We did 1 read at this point. So we should not return 0

File: src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+			if n != *exec_block {
+				return 0
```
Comment (tdelabro) [inline src/lib.rs:98]: We did 1 read at this point. So we should not return 0

File: src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+			}
+			loop {
+				if let Some((b, (mut sub, account))) = subs.pop() {
+					if b == *exec_block {
+						if let Some(val) = sub.remaining_payments {
+							if val == 0 {
+								subs.pop();
+								continue
+							}
```
Comment (tdelabro) [inline src/lib.rs:107]: Why are we using `pop` a second time ?
We already withdraw this subscription from the vec on line 101

File: src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+						if T::Currency::transfer(
+							&account,
+							&sub.beneficiary,
+							sub.amount,
+							ExistenceRequirement::AllowDeath,
+						)
+						.is_err()
+						{
+							continue
+						}
```
Comment (tdelabro) [inline src/lib.rs:118]: You are not puting it back into the vec. So this subscription will never be triggered again. Is it what we want ?
Shouldn't we set it on `freeze` status

We should decide how we articulate those two things. 
Maybe we can have an `unfreeze` method that will put it back onto the queue. wdyt ?

File: src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+	}
+
+	#[pallet::error]
+	pub enum Error<T> {
+		NoneValue,
+	}
+
+	#[pallet::hooks]
+	impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {
+		fn on_initialize(n: T::BlockNumber) -> Weight {
```
Comment (tdelabro) [inline src/lib.rs:91]: Put the whole content of this hook on a function that return a weight. And run some benchmarks on it making the size of the queue vary. Then you will be able to call the weight function in order to get the correct value to return at the end

Comment (tdelabro) [reply src/lib.rs:91]: Example: https://paritytech.github.io/substrate/monthly-2021-09+1/src/pallet_elections_phragmen/lib.rs.html#253

And at the end of `do_phragmen` it return `T::WeightInfo::election_phragmen(weight_candidates, weight_voters, weight_edges)`.

We want the same logic

File: src/lib.rs (hunk 2 above)
Comment (gdnathan) [reply src/lib.rs:96]: How do i evaluate how much weight is one read ? i added this so we would be able to benchmark it in the future, but rn i don't really know how to compute weight

File: src/lib.rs (hunk 5 above)
Comment (gdnathan) [reply src/lib.rs:118]: We said that, at the moment, we wont put status, to make things as simple as possible. So i think for the moment let's just "unsubscribe" the user if he cant pay anymore

File: src/lib.rs (hunk 6 above)
Comment (gdnathan) [reply src/lib.rs:91]: They actually do a read and return 0 in the hook

File: src/lib.rs [hunk 7]
```diff
@@ -0,0 +1,210 @@
...
+		}
+	}
+
+	impl<T: Config> Pallet<T> {
+		pub fn do_execute_subscriptions(n: T::BlockNumber) -> Weight {
+			let mut itterations = 0;
+
+			let mut subs = unwrap_or_return!(Subscriptions::<T>::get(), 0);
+			let tmp = subs.clone();
+			let (exec_block, (_, _)) = unwrap_or_return!(tmp.last(), 0);
```
Comment (tdelabro) [inline src/lib.rs:171]: why do this clone. You can get the value directly from `subs`

File: src/lib.rs [hunk 8]
```diff
@@ -0,0 +1,210 @@
...
+			//       delete from storage
+			Ok(())
+		}
+	}
+
+	impl<T: Config> Pallet<T> {
+		pub fn do_execute_subscriptions(n: T::BlockNumber) -> Weight {
+			let mut itterations = 0;
+
+			let mut subs = unwrap_or_return!(Subscriptions::<T>::get(), 0);
```
Comment (tdelabro) [inline src/lib.rs:169]: `subs` -> `subscriptions`

File: src/lib.rs [hunk 9]
```diff
@@ -0,0 +1,210 @@
...
+			// no -> signed by beneficiary ?
+			//       other is subscribed to beneficiary ?
+			//       delete from storage
+			Ok(())
+		}
+	}
+
+	impl<T: Config> Pallet<T> {
+		pub fn do_execute_subscriptions(n: T::BlockNumber) -> Weight {
+			let mut itterations = 0;
```
Comment (tdelabro) [inline src/lib.rs:167]: only one `t` in `iterations`

File: src/lib.rs [hunk 10]
```diff
@@ -0,0 +1,210 @@
...
+			//        delete from storage
+			// no -> signed by beneficiary ?
+			//       other is subscribed to beneficiary ?
+			//       delete from storage
+			Ok(())
+		}
+	}
+
+	impl<T: Config> Pallet<T> {
+		pub fn do_execute_subscriptions(n: T::BlockNumber) -> Weight {
```
Comment (tdelabro) [inline src/lib.rs:166]: `n` is a terrible variable name.
use `current_block_number`

File: src/lib.rs [hunk 11]
```diff
@@ -0,0 +1,210 @@
...
+			let mut itterations = 0;
+
+			let mut subs = unwrap_or_return!(Subscriptions::<T>::get(), 0);
+			let tmp = subs.clone();
+			let (exec_block, (_, _)) = unwrap_or_return!(tmp.last(), 0);
+			if n > *exec_block {
+				return 0
+			}
+			loop {
+				if let Some((block, (mut sub, account))) = subs.pop() {
```
Comment (tdelabro) [inline src/lib.rs:176]: `sub` should be `subscription`
`account` `subscriber`
`block` `block_number`

**Review by tdelabro — CHANGES_REQUESTED** [review]

File: src/lib.rs [hunk 12]
```diff
@@ -0,0 +1,197 @@
...
+			//       delete from storage
+			Ok(())
+		}
+	}
+
+	impl<T: Config> Pallet<T> {
+		pub fn do_execute_subscriptions(n: T::BlockNumber) -> Weight {
+			let mut itterations = 0;
+
+			let subs = unwrap_or_return!(Subscriptions::<T>::get(n), 0);
```
Comment (tdelabro) [inline src/lib.rs:168]: You just did a read, so don't return 0.
Return weight of a read

File: src/lib.rs [hunk 13]
```diff
@@ -0,0 +1,197 @@
...
+				}
+				sub.remaining_payments = sub.remaining_payments.map(|amount| amount - 1);
+				let new_block = n + sub.frequency;
+                let new_subs = Subscriptions::<T>::take(new_block).map(|mut subs| {
+                    subs.push((sub.clone(), account.clone()));
+                    subs
+                }).unwrap_or_else(|| vec!((sub, account)));
+                Subscriptions::<T>::insert(new_block, new_subs);
+			}
+			T::WeightInfo::do_execute_subscriptions(itterations)
```
Comment (tdelabro) [inline src/lib.rs:194]: We should remove the key we used (current_block) from the storage

File: src/lib.rs [hunk 14]
```diff
@@ -0,0 +1,197 @@
...
+				{
+					continue
+				}
+				sub.remaining_payments = sub.remaining_payments.map(|amount| amount - 1);
+				let new_block = n + sub.frequency;
+                let new_subs = Subscriptions::<T>::take(new_block).map(|mut subs| {
+                    subs.push((sub.clone(), account.clone()));
+                    subs
+                }).unwrap_or_else(|| vec!((sub, account)));
+                Subscriptions::<T>::insert(new_block, new_subs);
```
Comment (tdelabro) [inline src/lib.rs:192]: You don't have to take it. You can just read it. You are going to replace the value with an insert just after anyhow.

Comment (tdelabro) [inline src/lib.rs:192]: Doing this you will do a read and a write for each subscription.
You should rather store the value you updated in a on-memory cache, update this cache as many time as you want and and only write them back to storage at the end of the whole loop

**Review by tdelabro — CHANGES_REQUESTED** [review]

**Pull request closed without merging**